### PR TITLE
boot_serial: Replace cbor auto-generated code with zcbor functions

### PR DESCRIPTION
**Pull request by nordicjm**
Replaces the auto-generated decoding/encoding files with inline code
    for encoding/decoding cbor data structures, this adds the benefit of
    allowing the elements to be in any order and reduces code size. To
    accommodate this, zcbor_bulk has been imported from Zephyr.

### CODE REVIEW

**Comment by nordicjm** [comment]
CC @de-nordic

**Comment by de-nordic** [comment]
I have built the PR vs main (918da26a676d1bcbc7c83ce2de335ef569f97f9f).

Building with command:
```west build -d boot -b nrf52840dk_nrf52840 bootloader/mcuboot/boot/zephyr/ -DCONFIG_MCUBOOT_SERIAL=y -DCONFIG_UART_CONSOLE=n```

Gave following stats for this PR:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       42930 B        48 KB     87.34%
             RAM:       27264 B       256 KB     10.40%
        IDT_LIST:          0 GB         2 KB      0.00%
```
and main:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       42922 B        48 KB     87.33%
             RAM:       27264 B       256 KB     10.40%
        IDT_LIST:          0 GB         2 KB      0.00%
```
there is some increase in used flash, but this probably happens because there is support for "upgrade" flag added with this PR, which is not present in main. 

Building with echo SMP command enabled:
```west build -d boot -b nrf52840dk_nrf52840 bootloader/mcuboot/boot/zephyr/ -DCONFIG_MCUBOOT_SERIAL=y -DCONFIG_UART_CONSOLE=n -DCONFIG_BOOT_MGMT_ECHO=y```

Stats from this PR:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       43126 B        48 KB     87.74%
             RAM:       27264 B       256 KB     10.40%
        IDT_LIST:          0 GB         2 KB      0.00%
```

stats for main:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       43234 B        48 KB     87.96%
             RAM:       27264 B       256 KB     10.40%
        IDT_LIST:          0 GB         2 KB      0.00%
```

so there are more than 100 bytes saved in flash already.

File: boot/boot_serial/src/boot_serial.c [hunk 1]
```diff
@@ -396,7 +410,27 @@ bs_upload(char *buf, int len)
...
+    zcbor_state_t zsd[4];
+    zcbor_new_state(zsd, sizeof(zsd) / sizeof(zcbor_state_t), (uint8_t *)buf, len, 1);
+
+    struct zcbor_map_decode_key_val image_upload_decode[] = {
+        ZCBOR_MAP_DECODE_KEY_DECODER("image", zcbor_uint32_decode, &img_num),
+        ZCBOR_MAP_DECODE_KEY_DECODER("data", zcbor_bstr_decode, &img_chunk_data),
+        ZCBOR_MAP_DECODE_KEY_DECODER("len", zcbor_size_decode, &img_size_tmp),
+        ZCBOR_MAP_DECODE_KEY_DECODER("off", zcbor_size_decode, &img_chunk_off),
+        ZCBOR_MAP_DECODE_KEY_DECODER("sha", zcbor_bstr_decode, &sha),
+        ZCBOR_MAP_DECODE_KEY_DECODER("upgrade", zcbor_bool_decode, &upgrade)
```
Comment (de-nordic) [inline boot/boot_serial/src/boot_serial.c:422]: We should probably #ifdef this line as it will not be that useful and popular in serial recovery.

Comment (de-nordic) [reply boot/boot_serial/src/boot_serial.c:422]: > If it's not here, won't the validation fail if a client sends it with their request?

Validation by zcbor_map_decode_bulk? The thing only searches for defined keys, so if you pass it a key it does not know, it ignores it. This has been intentionally done to work with extending protocol from client side, while server side lags behind.

Comment (nordicjm) [reply boot/boot_serial/src/boot_serial.c:422]: Ah was under the impression it did validation, will remove.

Comment (nordicjm) [reply boot/boot_serial/src/boot_serial.c]: Removed sha and upgrade

**Comment by de-nordic** [comment]
@nordicjm Can you rebase?

**Comment by nordicjm** [comment]
> @nordicjm Can you rebase?

Needs https://github.com/mcu-tools/mcuboot/pull/1678 to be merged

**Comment by nordicjm** [comment]
@utzig @de-nordic @nvlsianpu @d3zd3z can you review this PR please?

**Review by de-nordic — APPROVED** [review]

**Pull request merged**